/quizizz, weekly quizzes, deep-linking for /tell and more

harshil21 opened the pull request:
### New bot features:

- [x]  `/quizizz` will send one quiz to the user from the internet.

- [x] The bot will not reply back to your message if the message you send is prepended or appended with `!r` when replying to a message from it.

- [x] The bot will send 5 quizzes with a time limit in the 12B group and then a show a leaderboard after the quiz is over.

- [x]  Bot will now send a message to the user with a deep linked url for if `/tell` was called in a group instead of a private chat.

#### Internal code changes:
- [x]  Logging is now properly implemented.

- [x]  morning_goodness() should not run after 11am

- [x] Update docs and README, and add comments and type hints.

- [x] Refactor `main.py`

##### For testing purposes:
- [x] Revert all testbot variables to shanibot.

- [x] Fix bar graph

#### Issues fixed:

- [x]  Closed #20 

- [x]  Closed #21 

- [x]  Closed #22 

- [x] Closed #23 

- [x] Closed #24 

- Fixed `ReplyKeyboard` not being removed after timing out in `/8ball`.

#### General changes:
- The `typing...` chat action is now called in the beginning, as it takes time to produce a response anyway.
- Added a license since it's important that we have one
- Increased conversation timeout for `/tell` and `/8ball` to 40 seconds.
----

- [x] Verify everything works.

harshil21 committed:
Initial take on sending quizzes

Bot sends quiz of 5 questions with time limit of one day in the group.

Also added /quizizz for sending one quiz question.

harshil21 committed:
Merge branch 'master' into quizizz-and-logging

harshil21 committed:
PEP-8

harshil21 committed:
Added the !r feature

Exclamation mark entities are not available unfortunately :(

harshil21 committed:
Added deep linking for /tell

Now links user to private chat if used in group.

harshil21 committed:
implement logging module wide

- Changed sleep time to 5 in `main.py` as per API docs.

- Minor code adjustments and improvements.

harshil21 committed:
Also make it not reply if msg ends with `!r`

harshil21 committed:
Attempts to improve quiz.py

- Added more logging statements in `main.py`.

- Use bot properties from context instead of `constants.py`.

- `quiz.py` is left in test configuration state. Run only that file - @unclesam79

harshil21 committed:
Closed #20

harshil21 committed:
added type hints, updated docs

harshil21 committed:
Good morning quotes will only be sent at or before 11am

harshil21 committed:
Closed #21 and fixed annotations in bar graph

Inline clips might still be a bit unreliable as the links seem to expire in under a minute

harshil21 commented:
Majority of the work is now done. Entering testing phase soon.

harshil21 committed:
Closed #22 #23 #24

Also added more reactions and gave reworked media().

Fixed a few typos

harshil21 committed:
main.py refactored

Did some code cleanup and made logging a helper function.

harshil21 committed:
Update README.md

harshil21 commented:
Am fast tracking this. Only a few bugs need to be squashed.

harshil21 committed:
Fixed bugs

harshil21 committed:
Fixed bug, prep for release

- Added timeout message for /8ball.

- Fixed ReplyKeyboard not being removed after timing out.

- Removed print statements and prepared for release.

harshil21 commented:
Everything is complete! As with all releases, make sure to commit latest `user_data` in here.

harshil21 commented:
@unclesam79 Make sure to get your [personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line#creating-a-token) from GitHub settings and make a file called `github_token.txt` in `creds` before running the main script. This is so inline clips work.

harshil21 committed:
add GPL license

harshil21 commented:
After review, please 'squash and merge' and not just 'merge'

harshil21 committed:
fix old data type in sql_table

harshil21 committed:
Optimized for loop

harshil21 committed:
Update requirements.txt

harshil21 committed:
update requirements and pep-8

harshil21 committed:
Merge branch 'quizizz-and-logging' of https://github.com/tmslads/ShanispeakBot into quizizz-and-logging

unclesam79 commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 sentence_tokenizer.pickle
 db.sqlite3-shm
 db.sqlite3-wal
+*idea
```
What's this?

harshil21 replied:
Uh pycharm config folder

unclesam79 commented on chatbot.py:
```diff
@@ -5,12 +5,15 @@
 shanisirbot = chatterbot.ChatBot('The Shani Sir Bot',
                                  storage_adapter='chatterbot.storage.SQLStorageAdapter',
                                  logic_adapters=['chatterbot.logic.BestMatch'],
-                                 response_selection_method=response_selection.get_first_response,  # or get_random_response or get_most_frequent_response
```
Why not leave them in?

harshil21 replied:
Um was exceeding 120 char limit

unclesam79 replied:
Don't remove important comments just to adhere to a stupid char limit!

harshil21 replied:
I'll move the comment on top

unclesam79 commented on chatbot.py:
```diff
@@ -5,12 +5,15 @@
-                                 response_selection_method=response_selection.get_first_response,  # or get_random_response or get_most_frequent_response
+                                 response_selection_method=response_selection.get_first_response,
                                  preprocessors=['chatterbot.preprocessors.clean_whitespace'],
-                                 read_only=False)  # Set to True to disable further learning from conversations the bot has
```
??

harshil21 replied:
Was not removed, fixing comment...

unclesam79 commented on chatbot.py:
```diff
@@ -5,12 +5,15 @@
+                                 read_only=False)  # Set to True to disable further learning from conversations the bot
 
+shanisirbot.initialize()  # Does any work that needs to be done before the chatbot can process responses.
+get_tags = shanisirbot.storage.tagger.get_bigram_pair_string
```
Good job moving it from main.py

unclesam79 commented on commands.py:
```diff
@@ -70,40 +92,74 @@ def helper(update, context):
     @staticmethod
-    def swear(update, context):
+    def swear(update: Update, context: CallbackContext) -> None:
+        """Sends a couple of swear words to the user along with a message saying not to use them."""
```
"a couple of" = 4

unclesam79 commented on convos/bday.py:
```diff
@@ -1,50 +1,49 @@
 
-def bday_add_or_update(update, context):  # INPUT
+def bday_add_or_update(update: Update, context: CallbackContext) -> int:  # INPUT
     """Changes or adds your birthday into our records."""
```
We are not talking to the user in docstrings

harshil21 replied:
Right changing

unclesam79 commented on convos/magic.py:
```diff
@@ -2,58 +2,65 @@
+        logger(message=f"{update.effective_user.first_name} used /8ball in {get_chat_name(update)}"
+                       f" and on {update.message.reply_to_message.from_user.first_name}'s message.")
+
+        actual_msg = update.message.reply_to_message.message_id  # Reply to the reply of the received message.
```
Comment is a bit (very) confusing

harshil21 replied:
Will change

unclesam79 commented on convos/settings_gui.py:
```diff
@@ -116,12 +113,12 @@ def setting_msg(update, swap: bool = False):
 
 
-def prob_message(update, kind: str, column: str) -> [None, str]:
+def prob_message(update, kind: str, column: str) -> Union[None, str]:
```
Does this mean it can either return None or a str?

harshil21 replied:
That's right

unclesam79 commented on convos/settings_gui.py:
```diff
@@ -190,7 +187,7 @@ def prob_updater(update, context):  # PROBABILITY
 
 
-def change_prob(update, context):  # UPDATED
+def change_prob(update: Update, _: CallbackContext) -> int:  # UPDATED
```
What is the _ for?

unclesam79 replied:
Is CallbackContext passed automatically to these functions?

harshil21 replied:
> What is the _ for?

Since `context` is not being used, setting it as _ saves some memory (very little).


> Is CallbackContext passed automatically to these functions?

Yep

unclesam79 commented on convos/start.py:
```diff
@@ -1,29 +1,54 @@
+                                 reply_markup=tell_markup)
+
+        logger(message=f"{first_name} just tried using /tell in a {chat.type}. "
+                       f"A message telling them to use it private was sent.")
```
*A message telling them to use it _in private chat_ was sent.

harshil21 committed:
fixed capitalization of 'I' in string

harshil21 committed:
fixed bug

harshil21 committed:
quiz only once a week

harshil21 committed:
fixed bug where morning msg was sent after 11am

unclesam79 commented on bot_funcs/morning_wisher.py:
```diff
@@ -0,0 +1,66 @@
+    if diff.days < 1:
+        return
+
+    if right_now >= afternoon:
```
Use an `or` statement instead of two `if` statements.

harshil21 replied:
Hm yea okay

unclesam79 commented on helpers/namer.py:
```diff
@@ -0,0 +1,23 @@
+def get_chat_name(update: Update) -> str:
+    """Helper function to get name of private/group chat."""
+
+    name = update.effective_chat.title
```
Put this in an `else` block after line 22

harshil21 replied:
No? `update.effective_chat.title` is None if it's a private chat

unclesam79 replied:
Yea nevermind I was high

unclesam79 commented on inline.py:
```diff
@@ -1,22 +1,32 @@
     query = update.inline_query.query
     if not query:
-        context.bot.answer_inline_query(update.inline_query.id, results[:50])  # Show first 49 clips if nothing is typed
+        # Show first 49 clips if nothing is typed-
```
Why'd you choose 49?
Also, won't this show 50 results?

harshil21 replied:
Will show only 49, indexing is from a to n-1.

> Why'd you choose 49?

Cause telegram can handle max 50 queries so wanted one less for good measure

unclesam79 commented on main.py:
```diff
@@ -1,31 +1,18 @@
-logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)
+from bot_funcs import media_reactor, morning_wisher, bday_wisher, conversation, delete_pin
+from commands import BotCommands as bc
+from constants import shanibot
```
`import group_ids` as well for when we need to send some message through the console

harshil21 replied:
Must develop data_browser.py so it can do that

unclesam79 commented on main.py:
```diff
@@ -457,32 +88,32 @@ def prettyprintview():
-updater.job_queue.run_repeating(morning_goodness, 86400, first=1)
+updater.job_queue.run_repeating(bday_wisher.wish, 86400, first=1)  # Runs every time script is started, and once a day.
+updater.job_queue.run_repeating(morning_wisher.morning_goodness, 86400, first=1)
+updater.job_queue.run_repeating(inline.get_clips, 60, first=1)  # Have to re-fetch clips since links expire
```
re-fetch every 60 seconds?? You sure this is the best way to do it? Probably isn't a good idea to send requests so often.

harshil21 replied:
Have to fetch this often as clips expire after that period

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+
+    starts = ["See I'm keeping one quizizz now okay. You have one day to finish. For boards ok. I want everyone to do "
+              "it that's it.", "I have kept one quizizz now. I expect something okay.",
+              "Because of the bad like you say situation I have kept this online quizizz now. Do fast okay.",
```
horrible use of like you say, like you say

harshil21 replied:
please change then

unclesam79 replied:
next PR

unclesam79 commented on online/quiz_scraper.py:
```diff
@@ -0,0 +1,72 @@
+
+def a_quiz() -> Union[Tuple[List[str], List[List[str]], List[int]], None]:
+    page = r.randint(1, 76)
+    logger(message=f"Quiz obtained from {page=}.", debug=True)
```
What does `debug=True` do?

harshil21 replied:
Uses logging.debug()

unclesam79 commented on online/quiz_scraper.py:
```diff
@@ -0,0 +1,72 @@
+# TODO: Improve the entire thing-
+
+
+def a_quiz() -> Union[Tuple[List[str], List[List[str]], List[int]], None]:
```
Could you make function name and purpose a little clearer?

harshil21 replied:
Okay

unclesam79 commented on online/quiz_scraper.py:
```diff
@@ -0,0 +1,72 @@
+
+        answer = result.find('span', class_='ans')  # Returns answer number
+        right_answer = int(answer.text.strip().replace('ANS: ', ''))
+        all_answers.append(right_answer - 1)
```
wat

harshil21 replied:
scraping stuff being filtered

unclesam79 commented on online/util.py:
```diff
@@ -1,32 +1,37 @@
+
+    page = requests.get(FACT_URL)
+    results = BeautifulSoup(page.content, 'html.parser', parse_only=SoupStrainer(id='z'))  # Get only z tags
+    facts_list = [str(results.contents[index].contents[0]) for index in range(len(results))]
```
???

harshil21 replied:
`SoupStrainer` gets only z tags (since that's the only thing we need), rest is scraping stuff.

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+from online import quiz_scraper
+
+quizzes = []
+cwd = os.getcwd()
```
Maybe we could add cwd to constants.py

harshil21 replied:
not a constant though is it?

unclesam79 replied:
It is for a particular run

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+        context.bot_data['last_quiz'] = right_now
+
+    diff = right_now - context.bot_data['last_quiz']
+    print(diff)
```
Make print statement more informative; something like
`print(f"Last quiz was sent {diff.days} days ago.")`

harshil21 replied:
Uh okay will make that logging

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+                                     type=Poll.QUIZ, correct_option_id=answer, is_closed=False)
+        quizzes.append(quiz)
+
+    logger(message=f"The 5 quizzes were just sent to 12B successfully.")
```
Ain't it just _one_ 5 question quiz?

harshil21 replied:
Technically, Telegram counts it as 5 separate quizzes

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+
+    context.job_queue.run_once(callback=timedout, when=60 * 60 * 10, context=[quizzes])
+    context.bot_data['last_quiz'] = right_now  # Save new time for last quiz
+    context.dispatcher.persistence.flush()
```
What does this do?

harshil21 replied:
flush() force saves the data(since ptb has trouble saving bot_data in jobs for some reason), timedout() will run after 10 hours(closes quizzes)

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+        context.bot_data['quizizz'][user.id] = {'answers_right': 0, 'questions_answered': 0, 'answers_wrong': 0}
+
+    # Update/add entries if changed-
+    guy = context.bot_data['quizizz'][user.id]
```
Change `guy` to `lad`.

harshil21 replied:
ahh

unclesam79 commented on quiz.py:
```diff
@@ -0,0 +1,344 @@
+        os.remove(jpg_name_path)  # Remove jpg file
+
+
+def add_image(name: str, x: float or int, y: int, offset: float, zoom: float = 0.20) -> AnnotationBbox:
```
y coord can't be float?

harshil21 replied:
Can be, but not used as float in module. Changing anyway

unclesam79 commented on updatedb.py:
```diff
@@ -1,34 +0,0 @@
-import os
```
keep updatedb.py
move it to a separate folder if you'd like
it's useful when making changes to db.sqlite3

harshil21 replied:
Yea so we can use db browser directly to make changes to both table and schema

harshil21 replied:
much better than coding

harshil21 replied:
And you can also run sql code directly in db browser if we still need to

harshil21 committed:
address review

harshil21 committed:
Fixed critical bug

Fixed bug where user's with no pp/hidden pp weren't assigned as nobody.png

Also added timeout to `get_file` as read timeout was happening

harshil21 committed:
send quiz to tms group instead of 12b

harshil21 committed:
moved quiz.py and commands.py to bot_funcs and added updatedb.py back

harshil21 committed:
fixed file paths

harshil21 committed:
Made certain parts of shanifier() more random, added latest user_data, etc

- Also increased convo timeout of /8ball and /tell to 40 seconds.

- Fixed bug where morning message was sent before 8am.

- Added missing docs and comments in `conversation.py`

- Added a little help message  in /8ball so users understand how it works.

harshil21 commented:
Everything is added and pr should be ready for merge rn.

unclesam79 reviewed (approved):
Bloody amazing how much work you put into it. Quizizz is actually here.

gg wp bigguy
